src/cmds/pbsdsh.c

```diff
@@ -95,6 +107,94 @@ bailout(int sig)
  * @return - Void
  *
  */
+
+
+int
+host_match(char *line, char *host)
+{
+	if (NULL == host ) return 0;
+	int len = strlen(line);
+	static char domain[PBS_MAXHOSTNAME + 1];
+	char fullhost[PBS_MAXHOSTNAME + 1];
+	static struct in_addr addr;
+	static int addrvalid = -1;
+
+	if (line[len - 1] == '\n')
+		line[len - 1] = '\0';
+
+	if (strcmp(line, host) == 0)
+		return 1;
+
+	if (addrvalid == -1) {
+		addr.s_addr = inet_addr(host);
+		addrvalid = (addr.s_addr == INADDR_NONE) ? 0 : 1;
+	}
+
+	if (addrvalid) { /* compare IP addresses */
+		struct hostent *hp = gethostbyname(line);
+		int i;
+
+		if (hp == NULL)
+			return 0;
+
+		for (i = 0; hp->h_addr_list[i]; i++) {
+			if (memcmp(&addr, hp->h_addr_list[i],
+				   hp->h_length) == 0)
+				return 1;
+		}
+		return 0;
+	}
+
+	if (domain[0] == '\0') {
+		if (getdomainname(domain, (sizeof(domain) - 1)) == -1) {
+			perror("getdomainname");
+			exit(255);
+		}
+		if (domain[0] == '\0') {
+			int i;
+			char *dot;
+
+			if (gethostname(domain, (sizeof(domain) - 1)) == -1) {
+				perror("getdomainname");
+				exit(255);
+			}
+			if (domain[0] == '\0')
+				return 0;
+			if ((dot = strchr(domain, '.')) == NULL)
+				return 0;
+			for (i = 0, dot++; *dot; i++, dot++)
+				domain[i] = *dot;
+			domain[i] = '\0';
+		}
+	}
+	pbs_strncpy(fullhost, line, sizeof(fullhost));
+	strcat(fullhost, ".");
+	strcat(fullhost, domain);
+
+	if (strcmp(fullhost, host) == 0)
+		return 1;
+
+	return 0;
+}
+
+
+int find_hostline(char *host)
+{
+	if (NULL == getenv("PBS_NODEFILE")) return -1;
+	FILE *fp = fopen(getenv("PBS_NODEFILE"),"r");
+	if (NULL == fp) return -1;
+	
+	char line[HOST_NAME_MAX], *cp;
+	int i;	
+	for (i = 0; (cp = fgets(line, sizeof(line), fp)) != NULL; i++) {
+                if (host_match(line, host))
+                        break;
+        }	
+	fclose(fp);		
+	return i;
```

Comment:
What if 'host' is not found in PBS_NODEFILE? What will 'i' be as value as return?

Reply:
So if host is not found, it would return the i-th line (last line) of the PBS_NODEFILE. So if there are 10 entries, it would return index 10. Now later in the code, it says:

onenode=find_hostline(targethost);
if (onenode < 0) {
	err = 1;
}
oneode would then be 10 which it think as valid. The only time find_hostline() returns < 0 if unable to open PBS_NODEFILE or the PBS_NODEFILE variable is not in the environment. So perhaps, check if i is == to # of lines in the file, and if so, return -1.

Reply:
As i said, its undefined behavior unless we define it first.  It makes sense for a sanity check, too. I'll take it